### PR TITLE
MSSQL Script - Fields defaults and Updating IAmAlive without ETag

### PR DESCRIPTION
**Pull request by shayhatsor**
I'm currently completing the MySQL ADO Invariant implementation in Orleans.
This includes creating a complete MySQL setup script as available in MSSQL and adding relevant tests.
While converting *CreateOrleansTables_SqlServer.sql* to MySQL I came across what IMHO is a bug.
In current implementation, the **UpdateIAmAlivetimeKey** script updates the **ETag** field.
The **IAmAliveTime** is a purely informative field and isn't a part of the membership protocol.
This will cause silos to unnecessarily fail periodically when trying to update the membership table.
My solution is to use the MSSQL 2000 implementation for all versions of MSSQL, with the minor change of not updating the **ETag** in **UpdateIAmAlivetimeKey**.
@veikkoeeva what do you think?

### CODE REVIEW

**Comment by shayhatsor** [comment]
update: MSSQL unit tests currently fail with this change

**Comment by gabikliot** [comment]
Correct.
**I am alive indeed should NOT update eTag (neither this row etag nor this table version row).**

It is OK to update the etag correctness-wise, but it will increase the contention on the table, and may thus have a negative effect when silos join/leave.
Also please make sure this operation does not change the TableVersion of the table.

https://github.com/dotnet/orleans/blob/master/src/Orleans/SystemTargetInterfaces/IMembershipTable.cs#L110

**Comment by veikkoeeva** [comment]
@shayhatsor, @gabikliot  Indeed, currently/in the old, both the SQL 2000 version and the newer ones [update the etag](https://github.com/dotnet/orleans/pull/985/files#diff-fb656146a37aaecbf42ba11153b597b7L785) in the `OrleansMembershipTable` for the given row. If this is a problem, it's OK to remove the update. The table version is not touched (see previous link). Table version is not even checked, should it be checked (I gather from the interface description that not)?

It looks like you removed table definition for newer editions than SQL Server 2000. A few things to note:
* In newer SQL Server editions that `DATETIME` is not recommended ('deprecated' might be [too strong a word](http://www.brentozar.com/dates-times-sql-server/)) and `NTEXT` definitely [is deprecated](https://msdn.microsoft.com/en-us/library/ms187993.aspx).
* I am not sure if support for SQL Server 2000 is warranted, I added it because I understood there is one company using Orleans with it, but at the same time, I wouldn't want to be a hostage of backwards compatibility.
* I wonder if we should add a note to the script or to the interface documentation where it reads that

> This operation should only update the IAmAlive collumn and not change other columns. [...]
> This operation may automatically update the eTag associated with the given silo row, but it does not have to. It can also leave the etag not changed ("dirty write").

that it adds for silo contention to update the etag. Looking at it more carefully now, I see there is a contradiction in that only IAmAlive column should be updated, but etag can be updated (and it might introduce contention).

@shayhatsor I think you will get what you want by changing the "newer than SQL Server 2000" to use `NEWID()` like SQL Server 2000. There script is longish due to SQL Server 2000 support. If support for that is dropped, it'd get shorter. For MySQL, you don't necessarily need to take care of different versions (and thus can drop optimization like things I did). I think you don't also need to provide the view, which is more or less *currently* to troubleshooting purposes. The plan is to add schema versioning to it, which would tell which version of the database it is and use that to both troubleshooting and update purposes later on.

I remember also that MySQL has a limit of 64 characters on constraint names and one is currently some characters longer in this SQL Server script. It would be nice if the names would be consistent across databases so that exception messages would would carry the same invariant names. I don't remember which one it was, but I think you'll bump into it sooner or later.

I also have a "half-made" (it's missing membership table updates, if I remember correctly) MySQL script on my computer, but we are moving currently and I already packed my desktop machine and shipped it to our new home. I might be able to retrieve on Sunday or Monday if it matters.

**Comment by veikkoeeva** [comment]
@shayhatsor Do you know the reason for the failure? Do the tests check etag (explicitly or implicitly) after ImAlive update?

**Comment by gabikliot** [comment]
@veikkoeeva , Table version is not even checked, indeed, that is correct.

There is no contradiction. "IAmAlive column should be updated, but etag can be updated , but does not have to". That is exactly the correct formulation. In Azure Table we have no way NOT to update the etag. It is updated automatically by the service every time the row is changed. But the MBR protocol does not require that etag to be updated in the IAmAlive case.
In SQL case it will be better not to update etag, if we can do that. But if we have to update it, its actually not a big deal, since this is per row (and not a global one, like Table version), thus the contention will not be significant.

So overall, in the SQL case, it is a pretty minor optimization I think. But it would not hurt to do it.

**Comment by veikkoeeva** [comment]
@gabikliot Oh, OK. I misread it. I'm of the opinion this optimization is a good thing also. The system is both better tuned and it exposes also nicely the fact that different implementations have some wiggle room. The protocol is robust like that.

**Comment by gabikliot** [comment]
Yes.

Clearly, I feel that @shayhatsor is an expert on Orlean's membership protocol now. Kudos!

**Comment by shayhatsor** [comment]
@veikkoeeva, i fixed the script following your suggestions. also, changed the ETag read to allow nulls - this is because of the difference between ROWVERSION which is never null, and BINARY(16) which can be null.
@gabikliot all relevant tests now pass successfully.

**Comment by shayhatsor** [comment]
@veikkoeeva, note that whenever you use *Convert.ToBase64String(selector.GetValueOrDefault<byte[]>("Something"))* you will get an exception if "Something" returns DBNull since default(byte[]) is null and Convert.ToBase64String(null) throws ArgumentNullException.

**Comment by veikkoeeva** [comment]
@shayhatsor Careful changes, look good to me! :+1:

**Comment by sergeybykov** [comment]
Is this ready to be merged?

**Comment by gabikliot** [comment]
Yes, from my perspective. Looks like @veikkoeeva approved too.

**Comment by veikkoeeva** [comment]
Yeah, look good to me. Sleeping over night, the only thing I have in my mind is that `VARBINARY` can be marked as `NOT NULL`, but I'm not sure if we should allow that possibility. One could think there cannot exist a row without some etag. I.e. there is no data to update before the first write and when it has happened, the subsequent updates just skip setting it.

**Comment by shayhatsor** [comment]
@veikkoeeva I'm not sure what you mean about `VARBINARY`. as of now, `VARBINARY` isn't used in the script. if you meant `BINARY`, then there's `NOT NULL` on it. we can add `DEFAULT NEWID()` to make the `INSERT` statements not set `ETag` explicitly. The `ETag` is always used for the *next* write, so setting it for the first row makes sense.

**Comment by veikkoeeva** [comment]
@shayhatsor I meant `BINARY(16)`, you were right, no `VARBINARY`. You are right, they should me marked as`NOT NULL DEFAULT NEWID()`? Like some of the other fields.

**Comment by shayhatsor** [comment]
@veikkoeeva I just found out [here](https://msdn.microsoft.com/en-us/library/ms182776.aspx) that:
>A nonnullable rowversion column is semantically equivalent to a binary(8) column. 

What I figure is that we should not use `NEWID()` - it's an overkill. Instead, use BINARY(8) as ETag and increment upon successful update (default will be 0). That will align with the behavior of `ROWVERSION`. What do you think ?

**Comment by veikkoeeva** [comment]
@shayhatsor I'm not sure what you propose. If we put the type of the column as `BINARY(8)` (additionally, `NOT NULL`), what you would you be using as the counter value and how? If we use something like `INT`, I think it's a bit tricky to implement, definitely a slower operation and lengthier SQL. Maybe what could be used is `DATETIME(4)` cast as `BINARY(8)`, but there's a chance to generate two similar etags (due to resolution).

It may not immediately clear from the code, as a clarification to the others who are reading this, but Orleans code reads the values as binary and before this change it'd read either the `UNIQUEIDENTIFIER` (*guid*) in SQL Server 2000 as binary or the `ROWVERSION` in newer editions as binary (though as null padded) and converted the byte array to base64 encoded string. So, from that standpoint, pretty much anything convertible to binary should work.

**Comment by shayhatsor** [comment]
@veikkoeeva I've force pushed a new version to show what I meant in the previous post.
I think this is much more readable and easier to port to other DBs.
all membership and reminders tests pass. (for both MSSQL 2000 and 2005+)
@sergeybykov - please don't merge yet as this is a change @veikkoeeva hasn't blessed yet.

**Comment by veikkoeeva** [comment]
@shayhatsor You know, I think you may be right about that increment by one approach. Also, the datetime function could be a faster one now that the resolution got dropped some time ago.

Now I'm worried about [this](https://github.com/dotnet/orleans/pull/985/files#diff-fb656146a37aaecbf42ba11153b597b7L1002) in that the locking ends to the semi-colon and consequently just before the IF branch has been evaluated, another thread has inserted. I *might* not be correct here. I'm seriously sleep-deprived due to the hackathon. :)

**Comment by shayhatsor** [comment]
first of all, good luck at the :godmode:`hackathon`:godmode:
I *think* I understand what you meant about the locking. I'm not sure it matters, since the primary key constraint will fail the insert anyway if two threads evaluate that they should insert.
Anyways, I'm not sure I understand the ETag mechanism used in the reminders. It's used only on deletes, it's not used on updates. maybe @gabikliot can shed some light.

**Comment by gabikliot** [comment]
Its a bit tricky, but fortunately I explained it already once.
https://github.com/dotnet/orleans/issues/637#issuecomment-124152405

**Comment by veikkoeeva** [comment]
@shayhatsor The failure shows up as exception in Orleans. One way could be to swallow exceptions that mention this specific constraint name, but I also think we should avoid unnecessary exceptions.

You likely already thought about this, but as a point to other readers, the return value is an opaque binary blob transformed to base64 encoded string by Orleans (and back to binary when writing to the database), the databases need not to have the same implementation but just to make sure they can provide atomically a value that differs from the previous one.

**Comment by shayhatsor** [comment]
@veikkoeeva I've checked what you said about the locking. It works because `WITH(updlock, holdlock)` creates the needed lock for the duration of the transaction, so we're good. please bless this PR :)
@gabikliot thanks for the explanation. Currently, `ReminderException` can be thrown:
1) [here](https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/ReminderService/LocalReminderService.cs#L162) - it seems to me that reminder upsert (in any implementation) should never fail for semantic reasons, so this will never happen when the reminders provider follows the protocol.
2) [here](https://github.com/dotnet/orleans/blob/master/src/OrleansRuntime/ReminderService/LocalReminderService.cs#L204) - why do we test the reminder etag upon deletes? shouldn't it also never fail for semantic reasons ? thanks for you help

**Comment by veikkoeeva** [comment]
@shayhatsor You might be right. It's what discussed [here](http://stackoverflow.com/questions/20192889/what-is-a-scope-of-a-table-hint-lock), and a question [here](http://stackoverflow.com/questions/8435177/trouble-with-tablockx-holdlock-scope). They are at best hints, so I believe there could be an exception anyhow. This is rather tricky thing as we can't be sure what isolation level the db is on. Though, we the constraint to be sure incorrect results can't sneak in.

Say, now that there are already some optimizations in, have you considered doing away hints like at  
https://msdn.microsoft.com/en-us/library/cc546505.aspx, from the
> This solution is very useful but works only if there is one row to be updated.

onwards? Just a thing. I think I agree with your changes. They are for the better. Maybe you could bless the code yourself too? ;)

**Comment by shayhatsor** [comment]
@sergeybykov this is ready for merge, a short description of the changes:
* Added `DEFAULT GETUTCTIME()` to every `TimeStamp DATETIME` field and removed from inserts.
* `ETag`
 * Changed all `BINARY(16)` to `BINARY(8) DEFAULT 0` and removed from inserts.
 * Removed the usage of `NEWID()`, instead we increment atomically
 * `OrleansMembershipTable`
   * Removed the update of `ETag` when only updating `IAmAliveTime`
    * In MSSQL 2005 - Changed `ROWVERSION` to `BINARY(8)`

**Comment by sergeybykov** [comment]
@shayhatsor Thanks. I'll merge it then.

**Comment by gabikliot** [comment]
@shayhatsor , re your question:

1) agree with the first case. upsert (in any implementation) should never fail for semantic reasons, so this will never happen when the reminders provider follows the protocol. So the case is redundant, but no harm in keeping it. Basically validates that indeed upsert returns a valid etag.

2) why we check etag upon reminder delete:

The reason is the semantics of reminder operations in the grain. `UpdateOrCreate` operates on the reminder string name and is unconditional with previous existence, and thus we don't validate etag. `Unregister` on the other hand operates on reminder instance. The API thus mandates to `GetReminder` and use the handle to unregister (or the one returned by Update). The intention was so that we always unregister only the reminder that we "own". We wanted to prevent "accidental" reminder deletes. This is especially important if there were 2 activations: one created the reminder for a first time, second updated (probably with same params, so basically like no op). Now imagine the first tries to delete. We don't want to allow that. We wanted instead the delete to fail and the 1st activation to be aware of the race. It may pick to Get and Delete after that: it is this activation choice how to handle the race. The 2nd activation will fail to delete now as well. and be aware of the race

So basically we decided to have `Update` unconditional and `Unregister` conditional, since we thought that would be the better application semantics. It did not have to be that ay. We can decide differently. The system will still work "correctly" but with different app semantics. But arguably, it would be harder for application code to reason about races with multiple activation w.r.t reminders.  As I said, it is not carved in stone, it can be changed and reminder concurrency semantics can be revisited.

**Comment by shayhatsor** [comment]
@gabikliot thanks for the explanation.
It seems to me that there are two options:
1) Checking ETag on every action. Then when a reminders race condition surfaces in a silo, it should commit suicide.
2) Don't check ETag on deletes. a grain that must be sure it doesn't have a duplicate activation will handle it itself, with whatever mechanism it desires. This is a well known limitation in Orleans.

IMHO, The current implementation is problematic:
The user is definitely not expecting a `ReminderException` which he can't do anything about but wait for a random amount of time for its local cache to refresh itself from the global reminders store. As you said, this exception might be thrown when there's a duplicate activation in order to protect the "owner" of the reminder. Correct me if I'm wrong, but a grain which is not the "owner" might still delete the reminder if its local reminders cache was just updated.

**Comment by gabikliot** [comment]
@shayhatsor , if you think the concurrency semantics of reminders should be changed, please open a new issue, describe the rational and the new proposal.

The current PR (#985) is unrelated, right?
We want first to make sure all semantics are implemented the way in all system store types (Azure, ZK, SQL) - which is, as far as I understand, this PR was about.
As a separate and unrelated step we can debate changing the semantics. In particular, @MSFT-AshleyIngram, who is working on #947, might be interested in this discussion.

**Comment by shayhatsor** [comment]
opened #1009
